Serverless Workflow Specification Sandbox Proposal

tsurdilo opened the pull request:
Signed-off-by: Tihomir Surdilovic <tsurdilo@redhat.com>

duglin commented on proposals/serverlessworkflow.md:
```diff
@@ -0,0 +1,170 @@
+
+TOC created the Serverless Working Group to investigate
+the Serverless landscape. The outputs of the WG included:
+- a [whitepaper]https://github.com/cncf/wg-serverless/blob/master/whitepapers/serverless-overview/README.md) that:
```
missing "(" before https.

tsurdilo replied:
fixed

duglin commented on proposals/serverlessworkflow.md:
```diff
@@ -0,0 +1,170 @@
+    around: Events, Deployments and Workflows
+  - provide additional education as needed
+
+In addition to CloudEvents specification, the Serverless WG has agreed to form a sub working group to work on
```
Since you didn't mention CE before this reference it might ease the transition a bit if you added something like:
```
The Working Group's first focus area was around making it easier for middleware to process Events, which led to the development of the [CloudEvents](https://cloudevents.io) specification. Additionally, the Serverless WG...
```

tsurdilo replied:
good point. added.

duglin commented on proposals/serverlessworkflow.md:
```diff
@@ -0,0 +1,170 @@
+- finalize release v0.1.
+- keep improving specification to v1.0.
+- define requirements for implementation compliance.
+- define requirements for Serverless Workflow API/SDK.
```
It might be good to include something about verification of the spec through implementations and testing, or something to that effect

tsurdilo replied:
added

duglin commented on proposals/serverlessworkflow.md:
```diff
@@ -0,0 +1,170 @@
+
+We would like to request a dedicated github repository for the Serverless Workflow specification.
+
+We would like to request a dedicated website.
```
you should probably get a dedicated mailing list too - if sandbox projects get those

tsurdilo replied:
added. yes good point

duglin commented on proposals/serverlessworkflow.md:
```diff
@@ -0,0 +1,170 @@
+
+Nokia, Red Hat, Camunda, Futurewei, Pivotal / VMware, IBM, Twilio, Codid
+
+And the following have voting rights (today), which means they regularly
```
Does this apply to workflow? I thought you guys had "maintainers" and those are the folks who can "vote".  I didn't think your voting rights were based on attendance like CE does it

tsurdilo replied:
you are correct, updated

tsurdilo commented:
Thanks for review @duglin !

tsurdilo commented:
Hi @amye can you let me know what to expect as far as time goes for moving this forward? Thanks!

amye commented:
This needs @cncf/toc review first for which SIG should be reviewing.

duglin commented:
@amye @lizrice can we get it added to an upcoming TOC call agenda?

amye commented:
We should be able to to get to this before the next call.

brendandburns commented:
@amye @lizrice 

We don't need a SIG review do we? Since this is a product of SIG-Serverless (and thus presumably would be reviewed by SIG-Serverless, but since SIG-Serverless sponsored the work, it seems redundant)

Anyway, regardless of that I'm willing to sponsor this one for inclusion in Sandbox.

fwiw, though Apache 2.0 doesn't work for specifications (this is a separate topic) but we need to discuss what the acceptable CNCF license is for specifications.

duglin commented:
Not the throw a monkey-wrench into it :-) but there is no SIG-Serverless.  There's a Serverless WG, and yes they approved it.  We've been trying to decide if we need a SIG-Serverless but haven't come to a conclusion yet

brendandburns commented:
Ah, right. I guess it's Sig-runtime then?

--brendan

________________________________
From: Doug Davis <notifications@github.com>
Sent: Thursday, March 12, 2020 3:35:09 PM
To: cncf/toc <toc@noreply.github.com>
Cc: Brendan Burns <bburns@microsoft.com>; Team mention <team_mention@noreply.github.com>
Subject: Re: [cncf/toc] Serverless Workflow Specification Sandbox Proposal (#376)


Not the throw a monkey-wrench into it :-) but there is no SIG-Serverless. There's a Serverless WG, and yes they approved it. We've been trying to decide if we need a SIG-Serverless but haven't come to a conclusion yet

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fcncf%2Ftoc%2Fpull%2F376%23issuecomment-598462006&data=02%7C01%7Cbburns%40microsoft.com%7Cee3bf3ba43dc450f587908d7c6d59f17%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637196493111356459&sdata=F5KKrnSE%2FDAn90MyIUJdA9957U4yM8WwtVeV%2B0KcHnM%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FABL4HATUVQOCZAUZZ2RI7DTRHFPR3ANCNFSM4LAR26GA&data=02%7C01%7Cbburns%40microsoft.com%7Cee3bf3ba43dc450f587908d7c6d59f17%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637196493111356459&sdata=uOtNo82e%2BTbtsUWkl4wwjoV1yJCdM%2F5t%2B2mcVXb5N00%3D&reserved=0>.

tsurdilo commented:
> Anyway, regardless of that I'm willing to sponsor this one for inclusion in Sandbox.

Thank you @brendandburns !!

tsurdilo commented:
updated to add links to more project pages such as roadmap doc, community contributions etc.

raravena80 commented:
@quinton-hoole and I had a brief conversation about this and we think this fits into SIG-App-Delivery better since it relates to how applications will behave/communicate when they get deployed as serverless. As opposed to runtime infrastructure that facilitates serverless workloads (i.e How you run a function within a container)

duglin commented:
I’m wondering if this is another example of something that doesn’t quite fit into an existing SIG (similar to how a lot us of don’t think CloudEvents does either), and therefore maybe we just need to bite the bullet and form SIG-Serverless.  We’ve been struggling, trying to decide where the serverless WG should go too. We were kind of hoping for a different name though, but didn’t really think of a good one.

/cc @markpeek @kenowens12

lizrice commented:
I'm also happy to sponsor this project to Sandbox. It's a fit in cloud native, and it feels very natural that it should be able to progress in the neutral space that the CNCF provides.

tsurdilo commented:
> I'm also happy to sponsor this project to Sandbox.

Thank you @lizrice !!

raravena80 commented:
Can we re-route this to SIG-App-Delivery? Objections? Comments? Thx.

duglin commented:
I was chatting with @lizrice and creating a SIG-Serverless seems like the best option and for this to then be under that SIG (along with CloudEvents).

However, it may take a bit of time to get the new SIG up and running.  A couple of things to note though:
- if my memory is correct, the serverless WG presented the workflow concept to the TOC a while ago back when we wanted to start the "workflow subgroup" under the serverless WG, so there was a bit of a review at that point
- the serverless wg approved the "workflow subgroup" asking to become a sandbox project prior to them opening this PR. So, if the serverless WG morphs into a SIG there's no reason to think it would change its mind

In order to not put any unnecessary delay into the processing of this PR since it's not their fault that we don't have the right SIG setup quite yet, what do people think about this:
- I can work with the other Serverless WG leads to write-up a proposed SIG-Serverless Charter to get that process started
- in the meantime, the Serverless WG can do what a SIG would normally do, and present this sandbox proposal to the TOC ( https://github.com/cncf/toc/blob/master/process/sandbox-process.png )

Thoughts?

raravena80 commented:
I also think that Serverless is a grey area, but imo the different components typically fall within the already defined SIGs. For example:

SIG-App-Delivery
- Serverless deployments
- Serverless CI/CD
- Serverless frameworks. i.e https://serverless.com/
- Serverless workflows/step functions 
- Serverless event specifications

SIG-Runtime
- Functions in containers
- Platforms that facilitate serverless/functions - Knative/Openfaas
- Sandboxes that can be used for serverless - Firecracker/Kata/gVisor
- Webassembly workers

SIG-Observability
- Serverless/function monitoring/tracing

SIG-Storage
- Serverless datastores

SIG-Network
- Serverless traffic tools

Perhaps the scope of SIG-Serverless could be(?):

- Serverless frameworks
- Serverless workflows/step functions 
- Serverless event specifications

caniszczyk commented:
+1 to a separate CNCF Serverless SIG, the WG was a prototype for how our current SIGs were structured and should be proposed as a SIG imho

xiang90 commented on proposals/serverlessworkflow.md:
```diff
@@ -0,0 +1,185 @@
+</p>
+
+You can see the Serverless Workflow specification examples
+[here](https://github.com/cncf/wg-serverless/blob/master/workflow/spec/spec-examples.md).
```
broken link?

tsurdilo replied:
Nice catch! we recently updated doc naming conventions so some links changed. I updated all the links as well as the path to the serverless workflow diagram included in the proposal. Let me know if you run into anything again. Thanks!

tsurdilo commented:
Hi @amye, could you give me an update on the overall status of this pr review? What kind of timeline can we expect at this point? Thanks!

amye commented:
SIG Runtime meets at 8am Pacific on the first and third Wednesdays of the month, https://docs.google.com/document/u/2/d/1k7VNetgbuDNyIs_87GLQRH2W5SLgjgOhB6pDyv89MYk/edit - suggest adding a presentation (when you're ready!) to their agenda.
They'll do a review in their issue queue: https://github.com/cncf/sig-runtime/issues

tsurdilo commented:
@amy Thanks for the info! I think we have been ready but just did't know that we had to add ourselves to the agenda :) Sorry. 
So just to make sure in https://docs.google.com/document/d/1k7VNetgbuDNyIs_87GLQRH2W5SLgjgOhB6pDyv89MYk/edit#

* could we do the presentation on the April 16th meeting? 
* Do I add the presentation link (I assume it can be a google docs link right?) under "Pending Presentations" or under "Notes:" in the "April 16th, 2020" entry?

Sorry for all the questions just never done this before :) Thanks again.

tsurdilo commented:
@amye ^

amye commented:
I would add into the Pending Presentations section and reach out to the chairs to help schedule.(@raravena80 is a good resource here.)
I already see a presentation on for April 16th, they may not want to do two presentations in one meeting.

raravena80 commented:
@tsurdilo assuming that after these comments in this PR [[1]](https://github.com/cncf/toc/pull/376#issuecomment-600342617) [[2]](https://github.com/cncf/toc/pull/376#issuecomment-601889624) the @cncf/toc (cc: @lizrice) ultimately decided this be reviewed by SIG-Runtime and not SIG-App-Delivery. You can open an issue [here](https://github.com/cncf/sig-runtime/issues) and feel free to add the presentation to any upcoming [SIG-Runtime](https://docs.google.com/document/d/1k7VNetgbuDNyIs_87GLQRH2W5SLgjgOhB6pDyv89MYk) meeting. We started added time estimates to the meetings so you can also check that the presentation fits within the agenda items.

Thanks!

tsurdilo commented:
@raravena80 @amye Thanks guys for your help! 
* Added Issue: https://github.com/cncf/sig-runtime/issues/19
* Added presentation to the May 7th, 2020 Agenda

Please let me know if there is anything else I need to do. Thanks!!

tsurdilo commented:
@raravena80 Hi, just a quick question, I'm not sure how long the sig-runtime meets are, and given that there is already a presentation scheduled for 20 minutes on April 16th, do you think there is time for pres for this proposal to be after it? April 16th is alot better date for me at the meet time. Please let me know what you think. Thanks.

raravena80 commented:
@tsurdilo the meetings run an hour. It depends on how much time you think it will take but It might be better to schedule it on May 7th. (We have another presentation on April 16h) Thx!

tsurdilo commented:
@raravena80 ok May 7th it is 👍 Thanks!

resouer commented:
@tsurdilo 

Based on @raravena80 's valuable https://github.com/cncf/toc/pull/376#issuecomment-601963968, specifically:

> SIG-App-Delivery
>
> Serverless deployments
> Serverless CI/CD
> Serverless frameworks. i.e https://serverless.com/
> Serverless workflows/step functions
> Serverless event specifications

vs

> SIG-Runtime
>
> Functions in containers
> Platforms that facilitate serverless/functions - Knative/Openfaas
> Sandboxes that can be used for serverless - Firecracker/Kata/gVisor
> Webassembly workers

As a workflow spec rather than workflow engine, it seems the core values of Serverless Workflow Specification project drops primarily in scope of SIG App Delivery and WG Serverless than SIG Runtime.

I'd suggest assign **SIG App Delivery + WG Serverless** as the review umbrella for this project, and kindly involve SIG Runtime etc in the review process.

/cc @brendandburns

tsurdilo commented:
@resouer Thanks for the info.
I am ok with whatever SIG this fits under best. We do not have a preference.
We have been sitting idle for a decision which SIG we fall under for over a month now, and would like to ask please for this decision to be finally made if possible :) 

Currently this PR is tagged with "sig-runtime" and if it is re-tagged under sig-app-delivery would you please provide us with a doc where we can add our presentation date/time as well as if opening a pr like we did for sig-runtime is necessary.

Help is much appreciated as I haven't gone through this process yet, and we would like to be able to present to a SIG and make progress to hopefully become an approved sandbox project as soon as possible. Thanks!

raravena80 commented:
AFAIK, currently [the TOC requests the SIGs to review](https://github.com/cncf/toc/blob/master/process/project_proposals.adoc#sandbox-process). I posted meeting information based on the assumption that the TOC had decided which SIG would ultimately review this project 😄. I think to avoid confusion it's probably best to have the TOC weigh in as to which SIG will ultimately review this project.

(Aside, it's SIG-Runtime's opinion that this should fall within the scope of SIG-App-Delivery).

amye commented:
Since the SIGs are in agreement here, I will move it. 
@tsurdilo: [SIG App-Delivery's](https://docs.google.com/document/d/1OykvqvhSG4AxEdmDMXilrupsX2n1qCSJUWwTc3I7AOs/edit#heading=h.n0ydhm3aurr) next meeting is April 15 - does that work for you?

tsurdilo commented:
@amye Yes, that would be great. Thanks! Do I need to put it on some agenda or create a pr?

amye commented:
@tsurdilo: see the link in previous comment, add yourself to the agenda for 4-15.

tsurdilo commented:
@brendandburns @lizrice Hi guys, thanks for stepping up and be willing to sponsor this specification. We had our toc pres April 15th and haven't heard the results yet. 
There were some requests to engage with different cncf projects during the presentation which we have followed up and are working on. Haven't heard back from the SIG regarding a decision.

The specification currently resides inside a github sub-directory of the cncf serverless group which makes it really hard to set up the proper infrastructure to engage other cncf projects even, and does now allows us to grow properly by being able to set up proper hooks, continuous integration testing etc. 

I would really like to as you guys as well as cncf and sig to please allow us to at least have our own github repository in order to be able to grow but as well be able to properly engage other cncf projects if that is requested at this point. Let me know if this is possible. Thanks!

amye commented:
@AloisReitbauer, @resouer, @bryanl - checking on documentation for SIG Review. Thanks!

tsurdilo commented on proposals/serverlessworkflow.md:
```diff
@@ -0,0 +1,185 @@
+
+Here is diagram that illustrates the Serverless Workflow functional flow:
+
+<p align="center">
```
Good point. Will try to create SVG from our spec images and use instead of png. Thanks!

tsurdilo commented:
@sdake thank you for the review and recommendations. Updated :)

tsurdilo commented:
@amye I think our entire group would really like a detailed explanation on the decision. After the review and some meetings I believe that we were told that with the incoming CNCF Sandbox process, we would not be granted the sandbox status for our project at this time.

We are not here to "complain" about any decisions, however it is important to us on the detailed reasons why, when:
* we have received project sponsors
* we are an active project 
* we have received a number of "lgtm" and support from the community
* we have followed all the sig recommendations and keep working hard to make sure what we were told during the project review is done asap and with best efforts

I have to be able to explain the team why the decision is made because we have a small but very active community of people that are interested in the growth of this project, and not being able to explain to them detailed reasons does not put the core contributors in a good light atm.

amye commented:
The SIGs provide written reviews to the TOC, but at this point in time, the existing sandbox process is the same. 
Can you point me to where you received different information?

tsurdilo commented:
@amye We had a follow up Slack meeting with Alois and Bryan on May 15th at which we received this information.

amye commented:
@bryanl, @AloisReitbauer -- your input please.

bryanl commented:
I'm thinking wires may be cross here, but that isn't the message Alois gave. The process is changing but never was it communicated that new projects were not being accepted.

tsurdilo commented:
@bryanl sorry if I caused any confusion, i did not mean to say that any decisions were based on any future changes in the process, but that this was a topic discussed during our meeting. 
I think all we are looking for is the document from the SIG regarding a decision so can communicate it with the team. That's all :)

tsurdilo commented:
@bryanl One thing that confused me the most regarding this whole process is:
* we had our sandbox presentation April 29 during which we got some recommendations
* we followed recommendations, set up meetings with cncf project teams, worked hard towards what you guys said during the pres
* we meet on May 15th to where we presented our progress towards meeting you recommendations you guys set forth, and then were told something to the like of "there is nothing we can do for you at this time" :) In addition we were told to keep working for the requirements of incubation cncf project.

This is all really confusing and I you cannot blame me for feeling that our project was not given a fighting chance from the start, no matter how much work (and trust me it was alot) we put in.

caniszczyk commented:
This was accepted into the Sandbox from today's TOC meeting to go over the sandbox backlog.

Project onboarding will be here: https://github.com/cncf/toc/issues/494
